File: modules/agent/src/serializers/comment.py
"""Serializers for comment-related models."""


def comment_to_response(comment) -> dict:
    """Convert Comment model to dictionary."""
    creator = comment.creator
    created_by_name = None
    created_by_email = None
    individual_id = None
    if creator:
        first = creator.first_name or ""
        last = creator.last_name or ""
        created_by_name = f"{first} {last}".strip() or None
        created_by_email = creator.email
        individual_id = creator.individual_id

    return {
        "id": comment.id,
        "tenant_id": comment.tenant_id,
        "commentable_type": comment.commentable_type,
        "commentable_id": comment.commentable_id,
        "content": comment.content,
        "created_by": comment.created_by,
        "created_by_name": created_by_name,
        "created_by_email": created_by_email,
        "individual_id": individual_id,
        "parent_comment_id": comment.parent_comment_id,
        "created_at": comment.created_at.isoformat() if comment.created_at else None,
        "updated_at": comment.updated_at.isoformat() if comment.updated_at else None,
    }

File: modules/agent/internal/serializers/comment.go
package serializers

import (
	"strings"
	"time"

	"agent/internal/models"
)

// CommentResponse represents a comment response
type CommentResponse struct {
	ID              int64     `json:"id"`
	TenantID        int64     `json:"tenant_id"`
	CommentableType string    `json:"commentable_type"`
	CommentableID   int64     `json:"commentable_id"`
	Content         string    `json:"content"`
	CreatedBy       int64     `json:"created_by"`
	CreatedByName   *string   `json:"created_by_name"`
	CreatedByEmail  *string   `json:"created_by_email"`
	IndividualID    *int64    `json:"individual_id"`
	ParentCommentID *int64    `json:"parent_comment_id"`
	CreatedAt       time.Time `json:"created_at"`
	UpdatedAt       time.Time `json:"updated_at"`
}

// CommentToResponse converts Comment model to response
func CommentToResponse(comment *models.Comment) CommentResponse {
	resp := CommentResponse{
		ID:              comment.ID,
		TenantID:        comment.TenantID,
		CommentableType: comment.CommentableType,
		CommentableID:   comment.CommentableID,
		Content:         comment.Content,
		CreatedBy:       comment.CreatedBy,
		ParentCommentID: comment.ParentCommentID,
		CreatedAt:       comment.CreatedAt,
		UpdatedAt:       comment.UpdatedAt,
	}

	resp.CreatedByName, resp.CreatedByEmail, resp.IndividualID = extractCreatorInfo(comment.Creator)

	return resp
}

func extractCreatorInfo(creator *models.User) (*string, *string, *int64) {
	if creator == nil {
		return nil, nil, nil
	}

	name := buildUserFullName(creator.FirstName, creator.LastName)
	return name, &creator.Email, creator.IndividualID
}

func buildUserFullName(firstName, lastName *string) *string {
	var parts []string

	if firstName != nil && *firstName != "" {
		parts = append(parts, *firstName)
	}
	if lastName != nil && *lastName != "" {
		parts = append(parts, *lastName)
	}

	if len(parts) == 0 {
		return nil
	}

	name := strings.Join(parts, " ")
	return &name
}

// CommentsToResponse converts a slice of Comment models to responses
func CommentsToResponse(comments []models.Comment) []CommentResponse {
	result := make([]CommentResponse, len(comments))
	for i := range comments {
		result[i] = CommentToResponse(&comments[i])
	}
	return result
}
